module/zfs/zfs_znode.c

```diff
@@ -1049,6 +1075,11 @@ zfs_xvattr_set(znode_t *zp, xvattr_t *xvap, dmu_tx_t *tx)
 		    zp->z_pflags, tx);
 		XVA_SET_RTN(xvap, XAT_SPARSE);
 	}
+	if (XVA_ISSET_REQ(xvap, XAT_PROJINHERIT)) {
+		ZFS_ATTR_SET(zp, ZFS_PROJINHERIT, xoap->xoa_projinherit,
+		    zp->z_pflags, tx);
```

Comment:
It seems not easy to resolve such issue.

Assume that we only allow to set PROJINHERIT flag on an empty directory, then all the descendents (must be new) will inherit the project ID from the parent directory. Now, the directory is NOT empty. Under such case, how can we change the project ID? To guarantee the semantics, we can change neither the parent's project ID nor the child's project ID. Unless the user move all the descendents to a new directory with new project ID. That is too inconvenient for the users.

If we allow to the set PROJINHERIT flag on a non-emtry directory, then we have to scan the whole directory and set project ID for all the existing descendents when set PROJINHERIT flag on the directory. From the implementation view, that is almost not acceptable.

Then I would to say that the PROJINHERIT flag is not strictly kernel enforce mechanism, it needs the user/administrator to use that properly. We just offer related framework/interface, if user/administrator want to use such mechanism, they need to follow some rules. But if someone breaks the rule(s), then from kernel view, it will/can NOT prevent.

@ahrens @behlendorf How do you think?

Thanks!